iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/pipe/PipeConsensus.java

```diff
@@ -146,33 +147,61 @@ private void initAndRecover() throws IOException {
       }
     } else {
       // asynchronously recover, retry logic is implemented at PipeConsensusImpl
-      CompletableFuture.runAsync(
-              () -> {
-                try (DirectoryStream<Path> stream = Files.newDirectoryStream(storageDir.toPath())) {
-                  for (Path path : stream) {
-                    ConsensusGroupId consensusGroupId =
-                        parsePeerFileName(path.getFileName().toString());
-                    PipeConsensusServerImpl consensus =
-                        new PipeConsensusServerImpl(
-                            new Peer(consensusGroupId, thisNodeId, thisNode),
-                            registry.apply(consensusGroupId),
-                            path.toString(),
-                            new ArrayList<>(),
-                            config,
-                            consensusPipeManager,
-                            syncClientManager);
-                    stateMachineMap.put(consensusGroupId, consensus);
-                    consensus.start(true);
-                  }
-                } catch (Exception e) {
-                  LOGGER.error("Failed to recover consensus from {}", storageDir, e);
-                }
-              })
-          .exceptionally(
-              e -> {
-                LOGGER.error("Failed to recover consensus from {}", storageDir, e);
-                return null;
-              });
+      CompletableFuture<Void> future =
+          CompletableFuture.runAsync(
+                  () -> {
+                    try (DirectoryStream<Path> stream =
+                        Files.newDirectoryStream(storageDir.toPath())) {
+                      for (Path path : stream) {
+                        ConsensusGroupId consensusGroupId =
+                            parsePeerFileName(path.getFileName().toString());
+                        PipeConsensusServerImpl consensus =
+                            new PipeConsensusServerImpl(
+                                new Peer(consensusGroupId, thisNodeId, thisNode),
+                                registry.apply(consensusGroupId),
+                                path.toString(),
+                                new ArrayList<>(),
+                                config,
+                                consensusPipeManager,
+                                syncClientManager);
+                        stateMachineMap.put(consensusGroupId, consensus);
+                        consensus.start(true);
```

Comment:
can we start them after resetpeerlist?

Reply:
Not sure. @Pengzna cc

Reply:
we must start them in compelableFuture to avoid start failure. may put `resetPeerListWithoutThrow.accept` into the completableFuture.